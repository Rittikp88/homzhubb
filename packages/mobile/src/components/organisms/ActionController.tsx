import React from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { RecordAssetActions } from '@homzhub/common/src/modules/recordAsset/actions';
import { RecordAssetSelectors } from '@homzhub/common/src/modules/recordAsset/selectors';
import LeaseTermController from '@homzhub/mobile/src/components/organisms/LeaseTermController';
import { SaleTermController } from '@homzhub/mobile/src/components/organisms/SaleTermController';
import { ManageTermController } from '@homzhub/mobile/src/components/organisms/ManageTermController';
import { Country } from '@homzhub/common/src/domain/models/CountryCode';
import { TypeOfPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { AssetGroupTypes } from '@homzhub/common/src/constants/AssetGroup';
import { FurnishingTypes } from '@homzhub/common/src/constants/Terms';
import { ILastVisitedStep } from '@homzhub/common/src/domain/models/LastVisitedStep';
import { IState } from '@homzhub/common/src/modules/interfaces';

interface IProps {
  typeOfPlan: TypeOfPlan;
  isSplitAsUnits: boolean;
  onNextStep: () => void;
  country: Country;
  assetGroupType: string;
  furnishing: FurnishingTypes;
  lastVisitedStep: ILastVisitedStep;
}

interface IDispatchProps {
  setTermId: (termId: number) => void;
  getMaintenanceUnits: () => void;
}

interface IStateProps {
  currentTermId: number;
  currentAssetId: number;
}

type Props = IStateProps & IDispatchProps & IProps;

// TODO (Aditya): Optimize, look at un-required props drilling
class ActionController extends React.PureComponent<Props, {}> {
  public componentDidMount = (): void => {
    const { getMaintenanceUnits, assetGroupType } = this.props;
    if (assetGroupType === AssetGroupTypes.COM) {
      getMaintenanceUnits();
    }
  };

  public render = (): React.ReactNode => {
    const { onNextStep } = this.props;
    const {
      typeOfPlan,
      currentTermId,
      setTermId,
      currentAssetId,
      country: { currencies },
      assetGroupType,
      lastVisitedStep,
      furnishing,
    } = this.props;
    return (
      <>
        {typeOfPlan === TypeOfPlan.SELL && (
          <SaleTermController
            currentAssetId={currentAssetId}
            assetGroupType={assetGroupType}
            currencyData={currencies[0]}
            onNextStep={this.onNextStep}
          />
        )}
        {typeOfPlan === TypeOfPlan.RENT && (
          <LeaseTermController
            currentAssetId={currentAssetId}
            currentTermId={currentTermId}
            assetGroupType={assetGroupType}
            setTermId={setTermId}
            furnishing={furnishing}
            lastVisitedStep={lastVisitedStep}
            currencyData={currencies[0]}
            onNextStep={onNextStep}
          />
        )}
        {typeOfPlan === TypeOfPlan.MANAGE && (
          <ManageTermController
            currentAssetId={currentAssetId}
            assetGroupType={assetGroupType}
            currencyData={currencies[0]}
            onNextStep={this.onNextStep}
          />
        )}
      </>
    );
  };

  private onNextStep = async (type: TypeOfPlan): Promise<void> => {
    const { onNextStep, lastVisitedStep, currentAssetId } = this.props;

    const last_visited_step = {
      ...lastVisitedStep,
      listing: {
        ...lastVisitedStep.listing,
        type,
        is_listing_created: true,
      },
    };
    await AssetRepository.updateAsset(currentAssetId, { last_visited_step });
    onNextStep();
  };
}

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { setTermId, getMaintenanceUnits } = RecordAssetActions;
  return bindActionCreators({ setTermId, getMaintenanceUnits }, dispatch);
};

const mapStateToProps = (state: IState): IStateProps => {
  const { getCurrentTermId, getCurrentAssetId } = RecordAssetSelectors;
  return {
    currentTermId: getCurrentTermId(state),
    currentAssetId: getCurrentAssetId(state),
  };
};

const HOC = connect(mapStateToProps, mapDispatchToProps)(ActionController);
export { HOC as ActionController };
