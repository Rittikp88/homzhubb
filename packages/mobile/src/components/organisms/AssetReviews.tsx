import React, { memo, useEffect, useState } from 'react';
import { View, StyleSheet } from 'react-native';
import { useTranslation } from 'react-i18next';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { theme } from '@homzhub/common/src/styles/theme';
import { EmptyState } from '@homzhub/common/src/components/atoms/EmptyState';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { AssetReviewCard } from '@homzhub/mobile/src/components/molecules/AssetReviewCard';
import { AssetReviewsSummary } from '@homzhub/mobile/src/components/molecules/AssetReviewsSummary';
import { AssetReview } from '@homzhub/common/src/domain/models/AssetReview';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { icons } from '@homzhub/common/src/assets/icon';

interface IProps {
  saleListingId: number | null;
  leaseListingId: number | null;
}

const AssetReviews = (props: IProps): React.ReactElement => {
  const { saleListingId, leaseListingId } = props;
  const [reviews, setReviews] = useState<AssetReview[]>([]);
  const [reviewSummary, setReviewSummary] = useState<AssetReview | null>(null);

  const { t } = useTranslation(LocaleConstants.namespacesKey.property);

  useEffect(() => {
    if (!saleListingId && !leaseListingId) {
      return;
    }

    try {
      AssetRepository.getListingReviews({
        ...(leaseListingId && { lease_listing: leaseListingId }),
        ...(saleListingId && { sale_listing: saleListingId }),
      }).then((response) => {
        setReviews(response);
      });
      AssetRepository.getListingReviewsSummary({
        ...(leaseListingId && { lease_listing: leaseListingId }),
        ...(saleListingId && { sale_listing: saleListingId }),
      }).then((response) => {
        setReviewSummary(response);
      });
    } catch (e) {
      AlertHelper.error({ message: t('common:genericErrorMessage') });
    }
  }, []);

  if (reviews.length <= 0 && !reviewSummary) {
    return <EmptyState title={t('property:noReview')} icon={icons.reviews} />;
  }

  return (
    <View style={styles.container}>
      {reviewSummary && <AssetReviewsSummary reviewSummary={reviewSummary} containerStyle={styles.content} />}
      {reviews.length > 0 && (
        <Label type="large" textType="semiBold" style={styles.popularWith}>
          {t('common:popularWith')}
        </Label>
      )}
      {reviews.map((review: AssetReview) => (
        <AssetReviewCard
          key={review.id}
          description={review.description}
          reviewedAt={review.postedAt}
          reviewedBy={review.reviewedBy}
          isReported={review.isReported}
          overallRating={review.rating}
          pillars={review.pillarRatings}
        />
      ))}
    </View>
  );
};

const memoizedComponent = memo(AssetReviews);
export { memoizedComponent as AssetReviews };

const styles = StyleSheet.create({
  container: {
    backgroundColor: theme.colors.white,
    minHeight: 200,
  },
  content: {
    marginHorizontal: 16,
  },
  popularWith: {
    color: theme.colors.darkTint3,
    margin: 16,
  },
});
