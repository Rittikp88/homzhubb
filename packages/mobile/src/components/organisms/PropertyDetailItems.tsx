import React from 'react';
import { View, StyleSheet, TextInput } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { theme } from '@homzhub/common/src/styles/theme';
import { SpaceAvailableTypes, ISpaceAvailable } from '@homzhub/common/src/domain/repositories/interfaces';
import { HorizontalPicker, IDropdownOption, Label, Text, Dropdown } from '@homzhub/common/src/components';
import { ButtonGroup } from '@homzhub/mobile/src/components/molecules/ButtonGroup';
import { IPropertyDetailsData, IPropertyTypes } from '@homzhub/common/src/domain/models/Property';

interface IPropertyDetailsItemsProps {
  data: IPropertyDetailsData[];
  areaUnits: IDropdownOption[];
  spaceAvailable: ISpaceAvailable;
  propertyGroupSelectedIndex: number;
  propertyGroupTypeSelectedIndex: number;
  carpetAreaError: boolean;
  onPropertyGroupChange: (index: string | number) => void;
  onPropertyGroupTypeChange: (index: string | number) => void;
  onSpaceAvailableValueChange: (item: IPropertyTypes, index: string | number) => void;
  onCommercialPropertyChange?: (type: string, value: string | number) => void;
}

type Props = IPropertyDetailsItemsProps & WithTranslation;

export class PropertyDetailsItems extends React.PureComponent<Props, {}> {
  public render(): React.ReactNode {
    const { data, propertyGroupSelectedIndex, propertyGroupTypeSelectedIndex } = this.props;
    return (
      <View style={styles.container}>
        <ButtonGroup<number>
          data={this.fetchButtonGroupData<IPropertyDetailsData>(data)}
          onItemSelect={this.onPropertyGroupSelect}
          selectedItem={propertyGroupSelectedIndex}
          textType="text"
          textSize="small"
          fontType="semiBold"
          containerStyle={styles.buttonGroupContainer}
          buttonItemStyle={styles.propertyTypeButton}
          testID="btngrpPropertyGroup"
        />
        <ButtonGroup<number>
          data={this.fetchButtonGroupData<IPropertyTypes>(data[propertyGroupSelectedIndex].asset_types)}
          onItemSelect={this.onPropertyGroupTypeSelect}
          selectedItem={propertyGroupTypeSelectedIndex}
          textType="label"
          textSize="regular"
          fontType="regular"
          containerStyle={styles.buttonGroupContainer}
          buttonItemStyle={styles.assetTypeButton}
          testID="btngrpPropertyGroupType"
        />
        <View style={styles.propertyContainer}>
          {this.renderSpaceAvailable()}
          {this.renderCarpetArea()}
        </View>
      </View>
    );
  }

  public renderSpaceAvailable = (): React.ReactElement => {
    const { t, data, propertyGroupSelectedIndex } = this.props;
    const spaceAvailableElements: Array<React.ReactElement> = [];
    data?.[propertyGroupSelectedIndex]?.space_types.forEach((space: IPropertyTypes, index: number) => {
      const key = data?.[propertyGroupSelectedIndex]?.name;
      const pickerValue = this.findSpaceTypeValue(space);
      const onValueChange = (value: string | number): void => {
        this.onHorizontalPickerValueChange(space, value);
      };
      spaceAvailableElements.push(
        <View style={styles.picker} key={`${key}-${index}`}>
          <Label type="large" textType="regular" style={styles.label}>
            {space.name}
          </Label>
          <HorizontalPicker key={`${key}-${index}`} onValueChange={onValueChange} value={pickerValue} testID="ftlist" />
        </View>
      );
    });
    return (
      <>
        <Text type="regular" textType="semiBold" style={styles.typeProperty}>
          {t('spaceAvailable')}
        </Text>
        {this.renderCommercialPickers()}
        {spaceAvailableElements}
      </>
    );
  };

  public renderCommercialPickers = (): React.ReactNode => {
    const {
      propertyGroupSelectedIndex,
      spaceAvailable: { totalFloors, floorNumber },
    } = this.props;
    if (propertyGroupSelectedIndex !== 1) {
      return null;
    }
    const commercialElements: Array<React.ReactNode> = [];
    const commercialSpaces = [
      {
        id: 1,
        name: SpaceAvailableTypes.FLOOR_NUMBER,
        value: floorNumber,
      },
      {
        id: 2,
        name: SpaceAvailableTypes.TOTAL_FLOORS,
        value: totalFloors,
      },
    ];
    commercialSpaces.forEach((space: any, index: number) => {
      const onValueChange = (value: string | number): void => {
        this.onHorizontalPickerValueChange(space, value);
      };
      commercialElements.push(
        <View style={styles.picker} key={index}>
          <Label type="large" textType="regular" style={styles.label}>
            {space.name}
          </Label>
          <HorizontalPicker key={index} onValueChange={onValueChange} value={space.value} testID="ftlist" />
        </View>
      );
    });
    return commercialElements;
  };

  public renderCarpetArea = (): React.ReactNode => {
    const {
      t,
      propertyGroupSelectedIndex,
      areaUnits,
      carpetAreaError,
      spaceAvailable: { carpetArea, areaUnit },
    } = this.props;
    if (propertyGroupSelectedIndex === 0 || areaUnits.length === 0) {
      return null;
    }
    const labelStyles = { ...theme.form.formLabel };
    const errorLabelStyles = { ...theme.form.formLabel, color: theme.colors.error };
    return (
      <View style={styles.formContainer}>
        <View style={styles.carpetArea}>
          <Label type="regular" textType="regular" style={carpetAreaError ? errorLabelStyles : labelStyles}>
            {t('carpetArea')}
          </Label>
          <TextInput
            key="carpetArea"
            value={carpetArea}
            keyboardType="number-pad"
            style={carpetAreaError ? styles.errorTextInput : styles.textInput}
            placeholder={t('enterCarpetArea')}
            onChangeText={this.onCarpetAreaChange}
            maxLength={5}
            testID="txtipCarpetArea"
          />
          {carpetAreaError && (
            <Label type="regular" style={styles.error}>
              {t('enterCarpetArea')}
            </Label>
          )}
        </View>
        <View style={styles.areaUnit}>
          <Label type="regular" textType="regular" style={labelStyles}>
            {t('areaUnit')}
          </Label>
          <Dropdown
            data={areaUnits}
            value={areaUnit}
            listTitle={t('selectAreaUnit')}
            onDonePress={this.onAreaUnitChange}
            iconSize={16}
            iconColor={theme.colors.darkTint7}
            containerStyle={styles.dropdownContainer}
            testID="dpnAreaUnit"
          />
        </View>
      </View>
    );
  };

  public onCarpetAreaChange = (value: string | number): void => {
    const { onCommercialPropertyChange } = this.props;
    if (onCommercialPropertyChange) {
      onCommercialPropertyChange('carpetArea', value);
    }
  };

  public onHorizontalPickerValueChange = (item: IPropertyTypes, index: string | number): void => {
    const { onSpaceAvailableValueChange } = this.props;
    onSpaceAvailableValueChange(item, index);
  };

  public onPropertyGroupSelect = (index: string | number): void => {
    const { onPropertyGroupChange } = this.props;
    onPropertyGroupChange(index);
  };

  public onPropertyGroupTypeSelect = (index: string | number): void => {
    const { onPropertyGroupTypeChange } = this.props;
    onPropertyGroupTypeChange(index);
  };

  public onAreaUnitChange = (value: string | number): void => {
    const { onCommercialPropertyChange } = this.props;
    if (onCommercialPropertyChange) {
      onCommercialPropertyChange('areaUnit', value);
    }
  };

  public findSpaceTypeValue = (space: IPropertyTypes): number => {
    const {
      spaceAvailable: { bedroom, bathroom, balcony },
    } = this.props;
    switch (space.name) {
      case SpaceAvailableTypes.BATHROOM:
        return bathroom;
      case SpaceAvailableTypes.BALCONY:
        return balcony;
      case SpaceAvailableTypes.BEDROOM:
        return bedroom;
      default:
        return 0;
    }
  };

  private fetchButtonGroupData = <T extends { name: string }>(data: T[]): { title: string; value: number }[] => {
    return data.map((item, index) => ({
      title: item.name,
      value: index,
    }));
  };
}

export default withTranslation(LocaleConstants.namespacesKey.property)(PropertyDetailsItems);
const styles = StyleSheet.create({
  container: {
    marginHorizontal: 16,
    maxWidth: theme.viewport.width,
  },
  typeProperty: {
    color: theme.colors.darkTint4,
    marginBottom: 15,
  },
  propertyContainer: {
    marginHorizontal: 10,
    marginTop: 0,
    flexDirection: 'column',
  },
  textInput: {
    textAlign: 'left',
    borderWidth: 1,
    borderRadius: 4,
    paddingHorizontal: 14,
    paddingVertical: 12,
    borderColor: theme.colors.disabled,
  },
  errorTextInput: {
    textAlign: 'left',
    borderWidth: 1,
    borderRadius: 4,
    paddingHorizontal: 14,
    paddingVertical: 12,
    ...theme.form.fieldError,
  },
  formContainer: {
    flexDirection: 'row',
    marginBottom: 20,
  },
  carpetArea: {
    flex: 0.5,
    marginRight: 20,
  },
  areaUnit: {
    flex: 0.5,
  },
  picker: {
    flex: 1,
    flexDirection: 'row',
    marginBottom: 10,
    marginTop: 10,
  },
  label: {
    flex: 0.8,
    alignSelf: 'center',
    color: theme.colors.darkTint4,
  },
  error: {
    color: theme.form.formErrorColor,
    marginTop: 3,
  },
  buttonGroupContainer: {
    marginBottom: 28,
  },
  propertyTypeButton: {
    paddingVertical: 12,
    paddingHorizontal: 12,
  },
  assetTypeButton: {
    marginBottom: 12,
  },
  dropdownContainer: {
    height: 54,
  },
});
