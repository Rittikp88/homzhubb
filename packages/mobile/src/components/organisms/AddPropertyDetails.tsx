import React, { ReactElement } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { theme } from '@homzhub/common/src/styles/theme';
import { Button, FormTextInput, FurnishingSelection, Text, WithShadowView } from '@homzhub/common/src/components';
import { AssetDescriptionForm } from '@homzhub/mobile/src/components/molecules/AssetDescriptionForm';
import { PropertySpaces } from '@homzhub/mobile/src/components/organisms/PropertySpaces';
import { AssetListingSection } from '@homzhub/mobile/src/components/HOC/AssetListingSection';
import { AssetDescriptionDropdownValues } from '@homzhub/common/src/domain/models/AssetDescriptionForm';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { ISpaceCount, SpaceType } from '@homzhub/common/src/domain/models/AssetGroup';
import { ILastVisitedStep } from '@homzhub/common/src/domain/models/LastVisitedStep';
import { IUpdateAssetParams } from '@homzhub/common/src/domain/repositories/interfaces';

interface IDescriptionForm {
  carpetArea?: number;
  areaUnit?: number;
  buildingGrade?: string;
  facing?: string;
  flooringType?: number;
  yearOfConstruction?: number;
  totalFloors?: number;
  onFloorNumber?: number;
}

interface IFurnishingForm {
  furnishingDetails?: string;
  furnishingType?: string;
}

interface IOwnProps extends WithTranslation {
  assetId: number;
  assetDetails: Asset | null;
  spaceTypes: SpaceType[];
  handleNextStep: () => void;
  lastVisitedStep: ILastVisitedStep;
}

interface IOwnState {
  spacesForm: ISpaceCount[];
  descriptionForm: IDescriptionForm;
  furnishingForm: IFurnishingForm;
  descriptionDropdownValues: AssetDescriptionDropdownValues | null;
}

class AddPropertyDetails extends React.PureComponent<IOwnProps, IOwnState> {
  constructor(props: IOwnProps) {
    super(props);
    const { assetDetails } = this.props;

    this.state = {
      spacesForm: [],
      descriptionForm: {
        carpetArea: (assetDetails && assetDetails.carpetArea) || undefined,
        areaUnit: (assetDetails && assetDetails.carpetAreaUnit && assetDetails.carpetAreaUnit.id) || 1,
        buildingGrade: '',
        facing: (assetDetails && assetDetails.facing) || undefined,
        flooringType: (assetDetails && assetDetails.floorType) || undefined,
        yearOfConstruction: (assetDetails && assetDetails.construction_Year) || undefined,
        totalFloors: (assetDetails && assetDetails.totalFloors) || 0,
        onFloorNumber: (assetDetails && assetDetails.floorNumber) || 0,
      },
      furnishingForm: {
        furnishingDetails: (assetDetails && assetDetails.furnishingDescription) || undefined,
        furnishingType: (assetDetails && assetDetails.furnishing) || undefined,
      },
      descriptionDropdownValues: null,
    };
  }

  public async componentDidMount(): Promise<void> {
    await this.getDescriptionDropdownValues();
  }

  public render(): ReactElement {
    const { spaceTypes, t } = this.props;
    const { descriptionForm, furnishingForm, descriptionDropdownValues } = this.state;

    return (
      <>
        <Formik
          onSubmit={this.onSubmit}
          initialValues={{ ...descriptionForm, ...furnishingForm }}
          validate={FormUtils.validate(this.formSchema)}
        >
          {(formProps: FormikProps<FormikValues>): React.ReactNode => {
            return (
              <>
                <View style={styles.containerStyle}>
                  <>
                    <Text style={styles.headingStyle} type="small">
                      {t('property:spacesText')}
                    </Text>
                    <PropertySpaces onChange={this.handleSpaceFormChange} spacesTypes={spaceTypes} />
                  </>

                  {descriptionDropdownValues && (
                    <AssetDescriptionForm dropDownOptions={descriptionDropdownValues} formProps={formProps} />
                  )}

                  {descriptionDropdownValues && this.renderFurnishingFields(formProps)}
                </View>

                <WithShadowView>
                  <Button
                    type="primary"
                    title={t('common:continue')}
                    containerStyle={styles.buttonStyle}
                    onPress={(): Promise<void> => this.onSubmit(formProps.values)}
                  />
                </WithShadowView>
              </>
            );
          }}
        </Formik>
      </>
    );
  }

  private renderFurnishingFields = (formProps: FormikProps<FormikValues>): ReactElement => {
    const { t } = this.props;
    const onFurnishingChange = (value: string): void => this.setFurnishingStatus(formProps, value);

    return (
      <AssetListingSection title={t('property:furnishing')} containerStyles={styles.furnishingStyle}>
        <>
          <FurnishingSelection
            titleHidden
            value={formProps.values.furnishingType}
            onFurnishingChange={onFurnishingChange}
          />
          <FormTextInput
            style={styles.furnishingFieldStyle}
            name="furnishingDetails"
            label={t('property:furnishingDetails')}
            maxLength={400}
            inputType="default"
            formProps={formProps}
          />
        </>
      </AssetListingSection>
    );
  };

  private onSubmit = async (values: FormikValues): Promise<void> => {
    const {
      areaUnit,
      carpetArea,
      facing,
      flooringType,
      furnishingDetails,
      onFloorNumber,
      totalFloors,
      yearOfConstruction,
      furnishingType,
    } = values;
    const { spacesForm } = this.state;
    const { handleNextStep, assetId, lastVisitedStep } = this.props;

    const sanitizedSpaces = spacesForm
      .filter((item) => item && item.description !== '')
      .map((space) => {
        return {
          space_type: space.space_type,
          count: space.count,
          ...(space.description && { description: space.description }),
        };
      });

    const payload: IUpdateAssetParams = {
      carpet_area: carpetArea,
      carpet_area_unit: areaUnit,
      facing,
      floor_type: flooringType,
      furnishing_description: furnishingDetails,
      floor_number: onFloorNumber,
      total_floors: totalFloors,
      construction_year: yearOfConstruction,
      furnishing: furnishingType,
      spaces: sanitizedSpaces,
      last_visited_step: {
        ...lastVisitedStep,
        asset_creation: {
          ...lastVisitedStep.asset_creation,
          is_details_done: true,
          total_step: 4,
        },
      },
    };

    try {
      await AssetRepository.updateAsset(assetId, payload);
      handleNextStep();
    } catch (e) {
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };

  private setFurnishingStatus = (formProps: FormikProps<FormikValues>, furnishingType: string): void => {
    formProps.setFieldValue('furnishingType', furnishingType);
  };

  private handleSpaceFormChange = (id: number, count: number, description?: string): void => {
    const { spacesForm } = this.state;

    spacesForm[id] = { space_type: id, count, description };
  };

  private formSchema = (): yup.ObjectSchema<IDescriptionForm & IFurnishingForm> => {
    return yup.object().shape({
      carpetArea: yup.number().optional(),
      areaUnit: yup.number().optional(),
      buildingGrade: yup.string().optional(),
      facing: yup.string().optional(),
      flooringType: yup.number().optional(),
      yearOfConstruction: yup.number().optional(),
      totalFloors: yup.number().optional(),
      onFloorNumber: yup.number().optional(),
      furnishingType: yup.string().optional(),
      furnishingDetails: yup.string().optional(),
    });
  };

  private getDescriptionDropdownValues = async (): Promise<void> => {
    try {
      const response: AssetDescriptionDropdownValues = await AssetRepository.getAssetDescriptionDropdownValues();

      this.setState({
        descriptionDropdownValues: response,
        furnishingForm: {
          furnishingType: response.furnishingStatus[0].name,
        },
      });
    } catch (e) {
      AlertHelper.error({ message: e.message });
    }
  };
}

const addPropertyDetails = withTranslation()(AddPropertyDetails);
export { addPropertyDetails as AddPropertyDetails };

const styles = StyleSheet.create({
  containerStyle: {
    paddingHorizontal: theme.layout.screenPadding,
  },
  buttonStyle: {
    flex: 0,
    margin: 16,
  },
  furnishingStyle: {
    marginTop: 16,
  },
  headingStyle: {
    marginTop: 16,
    paddingVertical: 16,
    paddingLeft: 16,
    backgroundColor: theme.colors.moreSeparator,
  },
  furnishingFieldStyle: {
    height: 85,
    textAlignVertical: 'top',
    paddingTop: 10,
  },
});
