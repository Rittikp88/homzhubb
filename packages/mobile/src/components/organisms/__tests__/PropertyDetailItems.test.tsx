import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PropertyAssetGroupData } from '@homzhub/common/src/mocks/PropertyDetails';
import { AreaUnit } from '@homzhub/common/src/mocks/AreaUnit';
import { PropertyDetailsItems } from '@homzhub/mobile/src/components/organisms/PropertyDetailItems';

const mock = jest.fn();

describe('Property Details Item Component', () => {
  let component: ShallowWrapper;
  let props: any;
  let spaceAvailable: any;

  beforeEach(() => {
    spaceAvailable = {
      bedroom: 1,
      bathroom: 1,
      balcony: 1,
      floorNumber: 1,
      totalFloors: 1,
      carpetArea: '1000',
      areaUnit: 'Sq.ft',
    };
    props = {
      areaUnits: AreaUnit,
      spaceAvailable,
      propertyGroupSelectedIndex: 0,
      propertyGroupTypeSelectedIndex: 0,
      carpetAreaError: false,
      onPropertyGroupChange: mock,
      onPropertyGroupTypeChange: mock,
      onSpaceAvailableValueChange: mock,
    };
  });

  afterEach(() => jest.clearAllMocks());

  it('should render property details items component with data', () => {
    props = { ...props, data: PropertyAssetGroupData };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should call onPropertyGroupSelect', () => {
    props = { ...props, data: PropertyAssetGroupData };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    component.find('[testID="btngrpPropertyGroup"]').prop('onValueChange')();
    expect(mock).toHaveBeenCalled();
  });

  it('should call onPropertyGroupTypeSelect', () => {
    props = { ...props, data: PropertyAssetGroupData };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    component.find('[testID="btngrpPropertyGroupType"]').prop('onItemSelect')();
    expect(mock).toHaveBeenCalled();
  });

  it('should call onCarpetareaChange', () => {
    props = { ...props, data: PropertyAssetGroupData, onCommercialPropertyChange: mock, propertyGroupSelectedIndex: 1 };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    component.find('[testID="txtipCarpetArea"]').prop('onChangeText')();
    expect(mock).toHaveBeenCalled();
  });

  it('should not call onCarpetareaChange', () => {
    props = { ...props, data: PropertyAssetGroupData, propertyGroupSelectedIndex: 1 };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    component.find('[testID="txtipCarpetArea"]').prop('onChangeText')();
    expect(mock).toHaveBeenCalledTimes(0);
  });

  it('should call onSpaceAvailableValueChange', () => {
    props = { ...props, data: PropertyAssetGroupData, propertyGroupSelectedIndex: 0 };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    component.find('[testID="ftlist"]').prop('onValueChange')();
    expect(mock).toHaveBeenCalled();
  });

  it('should call onSpaceAvailableValueChange', () => {
    props = { ...props, data: PropertyAssetGroupData, propertyGroupSelectedIndex: 1 };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    component.find('[testID="ftlist"]').at(0).prop('onValueChange')();
    expect(mock).toHaveBeenCalled();
  });

  it('should call onAreaUnitChange', () => {
    props = { ...props, data: PropertyAssetGroupData, onCommercialPropertyChange: mock, propertyGroupSelectedIndex: 1 };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    component.find('[testID="dpnAreaUnit"]').prop('onDonePress')();
    expect(mock).toHaveBeenCalled();
  });

  it('should not call onAreaUnitChange', () => {
    props = { ...props, data: PropertyAssetGroupData, propertyGroupSelectedIndex: 1 };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    component.find('[testID="dpnAreaUnit"]').prop('onDonePress')();
    expect(mock).toHaveBeenCalledTimes(0);
  });

  it('should return value for space type', () => {
    props = { ...props, data: PropertyAssetGroupData, propertyGroupSelectedIndex: 1 };
    component = shallow(<PropertyDetailsItems {...props} t={(key: string): string => key} />);
    // @ts-ignore
    const { findSpaceTypeValue } = component.instance();
    const bathroom = findSpaceTypeValue({ id: 1, name: 'Bathroom' });
    expect(bathroom).toStrictEqual(1);
    const balcony = findSpaceTypeValue({ id: 1, name: 'Balcony' });
    expect(balcony).toStrictEqual(1);
    const bedroom = findSpaceTypeValue({ id: 1, name: 'Bedroom' });
    expect(bedroom).toStrictEqual(1);
    const defaultValue = findSpaceTypeValue({ id: 1, name: 'Other' });
    expect(defaultValue).toStrictEqual(0);
  });
});
