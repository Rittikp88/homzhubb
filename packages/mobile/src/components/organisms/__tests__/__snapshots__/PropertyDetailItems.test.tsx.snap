// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property Details Item Component should render property details items component with data 1`] = `
<View
  style={
    Object {
      "marginHorizontal": 16,
      "maxWidth": 750,
    }
  }
>
  <ButtonGroup
    buttonItemStyle={
      Object {
        "paddingHorizontal": 12,
        "paddingVertical": 12,
      }
    }
    containerStyle={
      Object {
        "marginBottom": 28,
      }
    }
    data={
      Array [
        Object {
          "title": "Residential",
          "value": 0,
        },
        Object {
          "title": "Commercial",
          "value": 1,
        },
      ]
    }
    fontType="semiBold"
    onItemSelect={[Function]}
    selectedItem={0}
    testID="btngrpPropertyGroup"
    textSize="small"
    textType="text"
  />
  <ButtonGroup
    buttonItemStyle={
      Object {
        "marginBottom": 12,
      }
    }
    containerStyle={
      Object {
        "marginBottom": 28,
      }
    }
    data={
      Array [
        Object {
          "title": "Residential House",
          "value": 0,
        },
        Object {
          "title": "Apartment/Condo",
          "value": 1,
        },
        Object {
          "title": "Villa",
          "value": 2,
        },
        Object {
          "title": "Penthouse",
          "value": 3,
        },
        Object {
          "title": "Studio Apartment",
          "value": 4,
        },
        Object {
          "title": "Independent Floor",
          "value": 5,
        },
      ]
    }
    fontType="regular"
    onItemSelect={[Function]}
    selectedItem={0}
    testID="btngrpPropertyGroupType"
    textSize="regular"
    textType="label"
  />
  <View
    style={
      Object {
        "flexDirection": "column",
        "marginHorizontal": 10,
        "marginTop": 0,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#4C5F72",
          "marginBottom": 15,
        }
      }
      textType="semiBold"
      type="regular"
    >
      spaceAvailable
    </Text>
    <View
      key="Residential-0"
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    >
      <Label
        style={
          Object {
            "alignSelf": "center",
            "color": "#4C5F72",
            "flex": 0.8,
          }
        }
        textType="regular"
        type="large"
      >
        Room
      </Label>
      <HorizontalPicker
        key="Residential-0"
        onValueChange={[Function]}
        testID="ftlist"
        value={0}
      />
    </View>
  </View>
</View>
`;
